signer/single_signer_validator.go

```diff
@@ -1,42 +1,90 @@
 package signer
 
 import (
+	"fmt"
+	"os"
 	"sync"
 
-	tmcrypto "github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/privval"
 	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
-	tm "github.com/tendermint/tendermint/types"
 )
 
 var _ PrivValidator = &SingleSignerValidator{}
 
 // SingleSignerValidator guards access to an underlying PrivValidator by using mutexes
 // for each of the PrivValidator interface functions
 type SingleSignerValidator struct {
-	PrivValidator tm.PrivValidator
-	pvMutex       sync.Mutex
+	config     *RuntimeConfig
+	chainState map[string]*SingleSignerChainState
+}
+
+type SingleSignerChainState struct {
+	filePV  *privval.FilePV
+	pvMutex sync.Mutex
+}
+
+func NewSingleSignerValidator(config *RuntimeConfig) *SingleSignerValidator {
+	return &SingleSignerValidator{
+		config:     config,
+		chainState: make(map[string]*SingleSignerChainState),
+	}
 }
 
 // GetPubKey implements types.PrivValidator
-func (pv *SingleSignerValidator) GetPubKey() (tmcrypto.PubKey, error) {
-	pv.pvMutex.Lock()
-	defer pv.pvMutex.Unlock()
-	return pv.PrivValidator.GetPubKey()
+func (pv *SingleSignerValidator) GetPubKey() (crypto.PubKey, error) {
+	keyFile := pv.config.KeyFilePathSingleSigner()
+	filePV := privval.LoadFilePVEmptyState(keyFile, "")
+	return filePV.GetPubKey()
```

Comment:
Is there a reason we want to load all the information on each method invocation?

Or can we load it once (such as close to `main()`) and then populate a field in `SingleSignerValidator`?

Reply:
`GetPubKey` is called once by the sentry after initializing the connection. The sentry uses this to determine if the remote signer is a validator or not. I thought about checking `len(chainState) > 0` then using the pub key from the first one (pub key from all will be the same), but that would require doing this anyways when the chainID does not yet exist in the map (which will always be the case since GetPubKey is called before doing any specific chainID operations)

Reply:
I updated it to load the pub key during construction